### PR TITLE
[1213] suppress application alert recipient in provider serialiser if…

### PR DESCRIPTION
**Pull request by davidgisbey**
### Context
Previously, the provider serialiser added the application alert recipient if nil. This PR fixes that.

### Changes proposed in this pull request
- Added a conditional which only adds the application alert recipient if the email is not nil

### Guidance to review
- I'm feeling really unsure about the change i made to Jake's previous test. If i left the test for "handles the missing data gracefully"  as:
`expect(subject['contacts'].detect { |c| c['type'] == 'application_alert_recipient' }['email']).to be_nil`
rather than: 
`expect(subject['contacts'].detect { |c| c['type'] == 'application_alert_recipient' }).to be_nil`

it fails, as it returns an empty array.

Would also appreciate some input on possibly renaming the generate_provider_contacts, generate_contacts and application_alert recipient method names.

### CODE REVIEW

File: app/serializers/provider_serializer.rb [hunk 1]
```diff
@@ -103,4 +103,12 @@ def application_alert_recipient
       telephone: ''
       }]
   end
+
+  def generate_contacts
+    if object.ucas_preferences&.application_alert_email != nil
```
Comment (dankmitchell) [inline app/serializers/provider_serializer.rb]: Don't think we need the `!= nil` here. Can we check?

File: spec/serializers/provider_serializer_spec.rb [hunk 2]
```diff
@@ -57,7 +57,25 @@
     it { should eq provider.ucas_preferences.send_application_alerts_before_type_cast }
   end
 
-  context "when UCAS preferences are missing" do
+  describe 'application alert recipient' do
+    context 'if not nil the application alert recipient is serialized upon instantiation of the contact object' do
```
Comment (heidar) [inline spec/serializers/provider_serializer_spec.rb]: This doesn't feel like a context. Could we also reorganise this test to make it easier to follow?
```
context 'when set' do
  it 'does not serialise the application alert recipient' do
...
  end
end
```

File: spec/serializers/provider_serializer_spec.rb [hunk 3]
```diff
@@ -57,7 +57,25 @@
...
+      subject { serialize(provider)['contacts'].find { |c| c[:type] == 'application_alert_recipient' } }
```
Comment (heidar) [inline spec/serializers/provider_serializer_spec.rb]: It might be good to break this up a bit? Readability feels like it could be better perhaps.
```
let(:application_alert_recipient) do
  serialize(provider)['contacts'].find { |contact| contact[:type] == 'application_alert_recipient' }
end

subject { application_alert_recipient }
```

File: app/serializers/provider_serializer.rb (hunk 1 above)
Comment (davidgisbey) [reply app/serializers/provider_serializer.rb]: You're right. Thanks Dan.

File: spec/serializers/provider_serializer_spec.rb (hunk 3 above)
Comment (davidgisbey) [reply spec/serializers/provider_serializer_spec.rb]: Have actioned both your changes. Thanks for your feedback Heidar 👍

File: app/serializers/provider_serializer.rb [hunk 4]
```diff
@@ -41,7 +41,7 @@ class ProviderSerializer < ActiveModel::Serializer
              # see generate_provider_contacts and return_application_alert_recipient for logic.
 
   attribute :contacts do
-    generate_provider_contacts + application_alert_recipient
+    generate_contacts
```
Comment (misaka) [inline app/serializers/provider_serializer.rb]: Not sure it's worth having this as separate method at this point, unless it's used somewhere else?

File: spec/serializers/provider_serializer_spec.rb [hunk 5]
```diff
@@ -57,16 +57,44 @@
...
+      it { should_not include 'application_alert_recipient' }
+    end
+  end
+
   context "when UCAS preferences are missing" do
     before do
       provider.ucas_preferences.destroy
       provider.reload
     end
+# no need to test for the application as it will not have been instantiated into the contacts object due to being nil
```
Comment (misaka) [inline spec/serializers/provider_serializer_spec.rb]: minor tick: indentation

File: spec/requests/api/v1/provider_spec.rb [hunk 6]
```diff
@@ -80,7 +80,7 @@ def get_provider_codes_from_body(body)
         build(:ucas_preferences,
               type_of_gt12: :coming_or_not,
               send_application_alerts: :none,
-              application_alert_email: 'application_alert_recipient@acmeuniversity.education.uk')
+              application_alert_email: nil)
```
Comment (misaka) [inline spec/requests/api/v1/provider_spec.rb]: Why remove this from the test? Would it make sense to test this instead of the `nil` case?

Comment (davidgisbey) [reply spec/requests/api/v1/provider_spec.rb]: In this test two providers are parsed into JSON. Of these two, one has an application alert recipient and one doesn't. So it now tests both cases in one test.
